src/foam/u2/view/ScrollDAOView.js

```diff
@@ -341,9 +245,273 @@ foam.CLASS({
     }
   ],
 
+  classes: [
+    {
+      name: 'Row',
+      extends: 'foam.u2.Element',
+
+      documentation: `Recycled DOM rows. The "view" ViewSpec is used to manually
+          tear down and rebuild row contents whenever "data" changes. Scroll
+          performance is optimized by using an inner view with no dynamic
+          bindings so that the only DOM operation is the row completely
+          replacing its contents.`,
+
+      imports: [ 'rowFormatter' ],
+
+      axioms: [
+        foam.u2.CSS.create({
+          code: function CSS() {/*
+            ^ {
+              display: block;
+              contain: layout;
+              will-change: transform;
+              padding: 5px;
+              box-sizing: border-box;
+            }
+            ^ * {
+              background-color: #dddddd;
+            }
+        */}
+        })
+      ],
+
+      properties: [
+        [ 'nodeName', 'li' ],
+        {
+          name: 'data',
+          postSet: function(old, nu) {
+            if ( this.state !== this.LOADED ) return;
+            this.el().innerHTML = this.rowFormatter.format(nu);
```

Comment:
It may be possible to use U2 in a "formatter" like setup if we can't find a way to optimize U2 for your use case.  U2 elements have an output method, so you can use them to build an HTML string by first constructing the virtual DOM using U2 and then outputting where you need.  But that final output should be handled internally by the system like it currently is so there's no way for an external user to just provide their own HTML strings.

Although you should be getting the same performance as if you just added a slot to your U2 elements, and update that slot to a new element everytime you need to render a large chunk of HTML.  That should be roughly the same speed as just replace innerHTML, and if its not yet it should be able to be optimized to be nearly the same speed.